File: apps/host/internal/githubhost/pulls.go
package githubhost

import (
	"context"
	"errors"
	"strconv"
	"strings"

	pb "armadra.local/host/gen/armadra/v1"
	"armadra.local/host/internal/githubapi"
	"armadra.local/host/internal/storage"
)

// Pull requests (design §8). The merge path is the reason most of this file
// exists: a merge carries the exact head the reader saw, and the Host re-reads
// the pull request and its checks before sending anything.

// mergeMethods reads the strategies the repository actually allows. A method
// the repository has disabled is never offered, so the panel cannot present a
// button that is guaranteed to be refused.
func (s *Service) mergeMethods(ctx context.Context, client *githubapi.Client, ref *pb.GithubRepositoryRef) []pb.GithubMergeMethod {
	repository, _, err := client.Repository(ctx, ref, s.now().UnixMilli())
	if err != nil {
		return nil
	}
	return repository.AllowedMergeMethods
}

func (s *Service) ListPulls(ctx context.Context, caller Caller, request *pb.ListGithubPullsRequest) (*pb.ListGithubPullsResponse, error) {
	if err := s.authorize(caller, ScopeRead); err != nil {
		return nil, err
	}
	client, err := s.client(ctx)
	if err != nil {
		return nil, err
	}
	ref, err := s.repository(ctx, request.GetRepository())
	if err != nil {
		return nil, err
	}
	page, err := decodeCursor(request.GetAfterCursor())
	if err != nil {
		return nil, err
	}
	now := s.now().UnixMilli()
	allowed := s.mergeMethods(ctx, client, ref)
	pulls, response, err := client.Pulls(ctx, ref, request.GetFilter(), page, pageLimit(request.GetLimit()), allowed, now)
	if err != nil {
		return nil, s.translate(err)
	}
	return &pb.ListGithubPullsResponse{
		Pulls:            pulls,
		NextCursor:       encodeCursor(response.NextPage),
		HasMore:          response.NextPage >= 2,
		RateLimit:        rate(response.Rate),
		FromCache:        response.FromCache,
		ObservedAtUnixMs: now,
		PollIntervalMs:   PollIntervalMS,
	}, nil
}

// GetPull reads one pull request with its file summary, reviews, comments and
// the checks for its current head. The checks are always read for the head this
// response reports, so a summary can never describe a different commit.
func (s *Service) GetPull(ctx context.Context, caller Caller, request *pb.GetGithubPullRequest) (*pb.GetGithubPullResponse, error) {
	if err := s.authorize(caller, ScopeRead); err != nil {
		return nil, err
	}
	client, err := s.client(ctx)
	if err != nil {
		return nil, err
	}
	ref, err := s.repository(ctx, request.GetRepository())
	if err != nil {
		return nil, err
	}
	now := s.now().UnixMilli()
	allowed := s.mergeMethods(ctx, client, ref)
	pull, response, err := client.Pull(ctx, ref, request.GetNumber(), allowed, now)
	if err != nil {
		return nil, s.translate(err)
	}
	files, err := client.PullFiles(ctx, ref, pull.Number, maxFiles)
	if err != nil {
		return nil, s.translate(err)
	}
	reviews, err := client.PullReviews(ctx, ref, pull.Number, maxComments)
	if err != nil {
		return nil, s.translate(err)
	}
	reviewComments, err := client.PullReviewComments(ctx, ref, pull.Number, maxComments)
	if err != nil {
		return nil, s.translate(err)
	}
	comments, err := client.IssueComments(ctx, ref, pull.Number, maxComments)
	if err != nil {
		return nil, s.translate(err)
	}
	result := &pb.GetGithubPullResponse{
		Pull:           pull,
		Files:          files,
		Reviews:        reviews,
		ReviewComments: reviewComments,
		Comments:       comments,
		RateLimit:      rate(response.Rate),
		PollIntervalMs: PollIntervalMS,
	}
	if pull.HeadSha != "" {
		checks, err := client.Checks(ctx, ref, pull.HeadSha, now)
		if err != nil {
			return nil, s.translate(err)
		}
		result.Checks = checks
	}
	if result.References, err = s.referencesFor(ctx, caller.WorkspaceID, ref, pb.GithubReferenceKind_GITHUB_REFERENCE_KIND_PULL_REQUEST, pull.Number); err != nil {
		return nil, err
	}
	return result, nil
}

func (s *Service) GetChecks(ctx context.Context, caller Caller, request *pb.GetGithubChecksRequest) (*pb.GithubCheckSummary, error) {
	if err := s.authorize(caller, ScopeRead); err != nil {
		return nil, err
	}
	client, err := s.client(ctx)
	if err != nil {
		return nil, err
	}
	ref, err := s.repository(ctx, request.GetRepository())
	if err != nil {
		return nil, err
	}
	now := s.now().UnixMilli()
	pull, _, err := client.Pull(ctx, ref, request.GetNumber(), nil, now)
	if err != nil {
		return nil, s.translate(err)
	}
	if pull.HeadSha == "" {
		return nil, storage.ErrNotFound
	}
	checks, err := client.Checks(ctx, ref, pull.HeadSha, now)
	if err != nil {
		return nil, s.translate(err)
	}
	return checks, nil
}

// validRef accepts a branch name a request may name. It is deliberately
// stricter than git itself: nothing here needs a ref carrying a control
// character or a pattern character.
func validRef(value string) bool {
	if value == "" || len(value) > 255 || strings.HasPrefix(value, "-") || strings.Contains(value, "..") {
		return false
	}
	for _, r := range value {
		if r <= 0x20 || r == 0x7f || strings.ContainsRune("~^:?*[\\", r) {
			return false
		}
	}
	return true
}

// CreatePull checks that the head branch exists on the remote before asking for
// a pull request, so a branch that was never pushed produces a clear answer
// instead of an opaque remote refusal.
func (s *Service) CreatePull(ctx context.Context, caller Caller, request *pb.CreateGithubPullRequest) (*pb.GithubPullRequest, error) {
	if err := s.authorize(caller, ScopeWrite); err != nil {
		return nil, err
	}
	client, err := s.client(ctx)
	if err != nil {
		return nil, err
	}
	ref, err := s.repository(ctx, request.GetRepository())
	if err != nil {
		return nil, err
	}
	base, head := request.GetBaseRef(), request.GetHeadRef()
	if !validRef(base) || !validRef(head) || base == head {
		return nil, ErrInvalid
	}
	if request.GetTitle() == "" || len(request.GetTitle()) > 256 || len(request.GetBody()) > 65536 {
		return nil, ErrInvalid
	}
	remoteHead, err := client.Ref(ctx, ref, head)
	if err != nil {
		if githubapi.CodeOf(err) == githubapi.CodeNotFound {
			return nil, storage.ErrNotFound
		}
		return nil, s.translate(err)
	}
	// The caller passes the head it believes is pushed; a branch that moved
	// since then would open a pull request for the wrong commits.
	if expected := request.GetExpectedHeadSha(); expected != "" && expected != remoteHead {
		return nil, storage.ErrConflict
	}
	body := map[string]any{"title": request.GetTitle(), "base": base, "head": head}
	text := request.GetBody()
	if number := request.GetLinkedIssueNumber(); number > 0 {
		// The link is written as text the remote itself understands, and also
		// recorded locally as a reference below.
		if text != "" {
			text += "\n\n"
		}
		text += "Closes #" + strconv.FormatInt(number, 10)
	}
	if text != "" {
		body["body"] = text
	}
	if request.GetDraft() {
		body["draft"] = true
	}
	allowed := s.mergeMethods(ctx, client, ref)
	pull, err := client.CreatePull(ctx, ref, body, allowed, s.now().UnixMilli())
	if err != nil {
		return nil, s.translate(err)
	}
	return pull, nil
}

// SubmitReview posts one review anchored to the commit the reviewer read.
// Inline comments carry that commit, so an outdated draft is recognisable
// rather than silently attached to a line it no longer describes.
func (s *Service) SubmitReview(ctx context.Context, caller Caller, request *pb.SubmitGithubReviewRequest) (*pb.GithubReview, error) {
	if err := s.authorize(caller, ScopeWrite); err != nil {
		return nil, err
	}
	client, err := s.client(ctx)
	if err != nil {
		return nil, err
	}
	ref, err := s.repository(ctx, request.GetRepository())
	if err != nil {
		return nil, err
	}
	event := ""
	switch request.GetState() {
	case pb.GithubReviewState_GITHUB_REVIEW_STATE_APPROVED:
		event = "APPROVE"
	case pb.GithubReviewState_GITHUB_REVIEW_STATE_CHANGES_REQUESTED:
		event = "REQUEST_CHANGES"
	case pb.GithubReviewState_GITHUB_REVIEW_STATE_COMMENTED:
		event = "COMMENT"
	default:
		return nil, ErrInvalid
	}
	if event == "REQUEST_CHANGES" && strings.TrimSpace(request.GetBody()) == "" {
		return nil, ErrInvalid
	}
	if len(request.GetBody()) > 65536 || len(request.GetComments()) > 200 {
		return nil, ErrInvalid
	}
	body := map[string]any{"event": event}
	if request.GetCommitSha() != "" {
		body["commit_id"] = request.GetCommitSha()
	}
	if request.GetBody() != "" {
		body["body"] = request.GetBody()
	}
	comments := []map[string]any{}
	for _, draft := range request.GetComments() {
		if draft.GetPath() == "" || len(draft.GetPath()) > 4096 || draft.GetLine() <= 0 || strings.TrimSpace(draft.GetBody()) == "" {
			return nil, ErrInvalid
		}
		side := draft.GetSide()
		if side != "LEFT" && side != "RIGHT" {
			return nil, ErrInvalid
		}
		comments = append(comments, map[string]any{"path": draft.GetPath(), "line": draft.GetLine(), "side": side, "body": draft.GetBody()})
	}
	if len(comments) > 0 {
		body["comments"] = comments
	}
	review, err := client.CreateReview(ctx, ref, request.GetNumber(), body)
	if err != nil {
		return nil, s.translate(err)
	}
	return review, nil
}

// MergePull refuses unless the remote still holds exactly the head, and exactly
// the check rollup, that the caller said it displayed. A locally green screen
// is not a promise the remote will still accept the merge, so everything is
// re-read here rather than trusted.
func (s *Service) MergePull(ctx context.Context, caller Caller, request *pb.MergeGithubPullRequest) (*pb.MergeGithubPullResponse, error) {
	if err := s.authorize(caller, ScopeWrite); err != nil {
		return nil, err
	}
	client, err := s.client(ctx)
	if err != nil {
		return nil, err
	}
	ref, err := s.repository(ctx, request.GetRepository())
	if err != nil {
		return nil, err
	}
	method, err := mergeMethodName(request.GetMethod())
	if err != nil {
		return nil, err
	}
	if request.GetExpectedHeadSha() == "" {
		return nil, ErrInvalid
	}
	now := s.now().UnixMilli()
	allowed := s.mergeMethods(ctx, client, ref)
	pull, _, err := client.Pull(ctx, ref, request.GetNumber(), allowed, now)
	if err != nil {
		return nil, s.translate(err)
	}
	refuse := func(reason string, checks *pb.GithubCheckSummary) *pb.MergeGithubPullResponse {
		return &pb.MergeGithubPullResponse{Merged: false, ReasonCode: reason, Pull: pull, Checks: checks}
	}
	if pull.HeadSha != request.GetExpectedHeadSha() {
		return refuse("HEAD_MOVED", nil), nil
	}
	if len(allowed) > 0 && !allows(allowed, request.GetMethod()) {
		return refuse("METHOD_NOT_ALLOWED", nil), nil
	}
	var checks *pb.GithubCheckSummary
	if request.GetExpectedCheckRollup() != pb.GithubCheckConclusion_GITHUB_CHECK_CONCLUSION_UNSPECIFIED {
		if checks, err = client.Checks(ctx, ref, pull.HeadSha, now); err != nil {
			return nil, s.translate(err)
		}
		if checks.Rollup != request.GetExpectedCheckRollup() {
			return refuse("CHECKS_CHANGED", checks), nil
		}
	}
	switch pull.Mergeable {
	case pb.GithubMergeableState_GITHUB_MERGEABLE_STATE_CONFLICTING:
		return refuse("NOT_MERGEABLE", checks), nil
	case pb.GithubMergeableState_GITHUB_MERGEABLE_STATE_BLOCKED:
		// Protection rules and merge queues are the remote's decision; the Host
		// reports what it was told rather than trying anyway.
		return refuse("BLOCKED", checks), nil
	}
	sha, err := client.Merge(ctx, ref, pull.Number, pull.HeadSha, method, request.GetCommitTitle(), request.GetCommitMessage())
	if err != nil {
		translated := s.translate(err)
		switch {
		case errors.Is(translated, ErrUnknownOutcome):
			// The merge may have been accepted. Re-reading is the only honest
			// answer; retrying would risk a second merge commit.
			if latest, _, readErr := client.Pull(ctx, ref, pull.Number, allowed, s.now().UnixMilli()); readErr == nil {
				pull = latest
				if latest.State == pb.GithubPullState_GITHUB_PULL_STATE_MERGED {
					return &pb.MergeGithubPullResponse{Merged: true, MergeSha: latest.HeadSha, Pull: latest, Checks: checks}, nil
				}
			}
			return refuse("UNKNOWN_OUTCOME", checks), nil
		case errors.Is(translated, storage.ErrConflict), errors.Is(translated, ErrInvalid):
			// The remote refused. Which refusal it was is decided by evidence:
			// a head that still matches means the merge itself was rejected,
			// not that the branch moved underneath the reader.
			if latest, _, readErr := client.Pull(ctx, ref, pull.Number, allowed, s.now().UnixMilli()); readErr == nil {
				pull = latest
				if latest.HeadSha != request.GetExpectedHeadSha() {
					return refuse("HEAD_MOVED", checks), nil
				}
			}
			return refuse("NOT_MERGEABLE", checks), nil
		case errors.Is(translated, ErrPermission):
			return refuse("BLOCKED", checks), nil
		}
		return nil, translated
	}
	latest, _, readErr := client.Pull(ctx, ref, pull.Number, allowed, s.now().UnixMilli())
	if readErr == nil {
		pull = latest
	}
	return &pb.MergeGithubPullResponse{Merged: true, MergeSha: sha, Pull: pull, Checks: checks}, nil
}

func allows(methods []pb.GithubMergeMethod, method pb.GithubMergeMethod) bool {
	for _, candidate := range methods {
		if candidate == method {
			return true
		}
	}
	return false
}

func mergeMethodName(method pb.GithubMergeMethod) (string, error) {
	switch method {
	case pb.GithubMergeMethod_GITHUB_MERGE_METHOD_MERGE:
		return "merge", nil
	case pb.GithubMergeMethod_GITHUB_MERGE_METHOD_SQUASH:
		return "squash", nil
	case pb.GithubMergeMethod_GITHUB_MERGE_METHOD_REBASE:
		return "rebase", nil
	}
	return "", ErrInvalid
}

File: apps/desktop/src/core/github/pulls.ts
/**
 * Pull request（设计 §8）。移植自 `apps/host/internal/githubhost/pulls.go`。
 *
 * 合并那条路径是这个文件大部分内容存在的原因：一次合并带着读者看到的那个确切
 * head，而 core 在发出任何东西之前会把 PR 和它的检查都重新读一遍。
 */

import {
  GetGithubPullResponseSchema,
  GithubCheckConclusion,
  GithubMergeMethod,
  GithubMergeableState,
  GithubPullState,
  GithubReferenceKind,
  GithubReviewState,
  ListGithubPullsResponseSchema,
  MergeGithubPullResponseSchema,
  create,
  type CreateGithubPullRequest,
  type GetGithubChecksRequest,
  type GetGithubPullRequest,
  type GetGithubPullResponse,
  type GithubCheckSummary,
  type GithubPullRequest,
  type GithubRepositoryRef,
  type GithubReview,
  type ListGithubPullsRequest,
  type ListGithubPullsResponse,
  type MergeGithubPullRequest,
  type MergeGithubPullResponse,
  type SubmitGithubReviewRequest,
} from "@armadra/protocol";

import type { GithubClient } from "./client";
import * as api from "./endpoints";
import { codeOf, githubError, isGithubError } from "./errors";
import { referencesFor } from "./references";
import {
  MAX_COMMENTS,
  MAX_FILES,
  POLL_INTERVAL_MS,
  SCOPE_READ,
  SCOPE_WRITE,
  decodeCursor,
  encodeCursor,
  pageLimit,
  rateLimit,
  type Caller,
  type GithubService,
} from "./service";

/**
 * 读仓库实际允许的策略。一个仓库已经禁用的策略永远不被提供，所以面板没法摆出一个
 * 注定被拒的按钮。
 */
async function mergeMethods(
  service: GithubService,
  client: GithubClient,
  ref: GithubRepositoryRef,
): Promise<GithubMergeMethod[]> {
  try {
    const result = await api.repository(client, ref, service.now());
    return [...result.repository.allowedMergeMethods];
  } catch {
    return [];
  }
}

export async function listPulls(
  service: GithubService,
  caller: Caller,
  request: ListGithubPullsRequest,
): Promise<ListGithubPullsResponse> {
  service.authorize(caller, SCOPE_READ);
  const client = service.client();
  const ref = service.repository(request.repository);
  const page = decodeCursor(request.afterCursor);
  const now = service.now();
  const allowed = await mergeMethods(service, client, ref);
  try {
    const result = await api.pulls(
      client,
      ref,
      request.filter,
      page,
      pageLimit(Number(request.limit)),
      allowed,
      now,
    );
    service.credentials.noteSuccess();
    return create(ListGithubPullsResponseSchema, {
      pulls: result.pulls,
      nextCursor: encodeCursor(result.response.nextPage),
      hasMore: result.response.nextPage >= 2,
      rateLimit: rateLimit(result.response.rate),
      fromCache: result.response.fromCache,
      observedAtUnixMs: BigInt(now),
      pollIntervalMs: BigInt(POLL_INTERVAL_MS),
    });
  } catch (error) {
    throw service.translate(error);
  }
}

/**
 * 读一条 PR，连同它的文件概要、评审、评论和**当前 head 的**检查。检查永远是对着
 * 这次响应报出来的那个 head 读的，所以一份概要不可能描述另一个提交。
 */
export async function getPull(
  service: GithubService,
  caller: Caller,
  request: GetGithubPullRequest,
): Promise<GetGithubPullResponse> {
  service.authorize(caller, SCOPE_READ);
  const client = service.client();
  const ref = service.repository(request.repository);
  const now = service.now();
  const allowed = await mergeMethods(service, client, ref);
  try {
    const read = await api.pull(client, ref, request.number, allowed, now);
    const files = await api.pullFiles(
      client,
      ref,
      read.pull.number,
      MAX_FILES,
    );
    const reviews = await api.pullReviews(
      client,
      ref,
      read.pull.number,
      MAX_COMMENTS,
    );
    const reviewComments = await api.pullReviewComments(
      client,
      ref,
      read.pull.number,
      MAX_COMMENTS,
    );
    const comments = await api.issueComments(
      client,
      ref,
      read.pull.number,
      MAX_COMMENTS,
    );
    const result = create(GetGithubPullResponseSchema, {
      pull: read.pull,
      files,
      reviews,
      reviewComments,
      comments,
      rateLimit: rateLimit(read.response.rate),
      pollIntervalMs: BigInt(POLL_INTERVAL_MS),
      references: referencesFor(
        service,
        caller.workspaceId,
        ref,
        GithubReferenceKind.PULL_REQUEST,
        read.pull.number,
      ),
    });
    if (read.pull.headSha !== "") {
      result.checks = await api.checks(client, ref, read.pull.headSha, now);
    }
    service.credentials.noteSuccess();
    return result;
  } catch (error) {
    throw service.translate(error);
  }
}

export async function getChecks(
  service: GithubService,
  caller: Caller,
  request: GetGithubChecksRequest,
): Promise<GithubCheckSummary> {
  service.authorize(caller, SCOPE_READ);
  const client = service.client();
  const ref = service.repository(request.repository);
  const now = service.now();
  let read;
  try {
    read = await api.pull(client, ref, request.number, [], now);
    service.credentials.noteSuccess();
  } catch (error) {
    throw service.translate(error);
  }
  if (read.pull.headSha === "") throw githubError("notFound");
  try {
    return await api.checks(client, ref, read.pull.headSha, now);
  } catch (error) {
    throw service.translate(error);
  }
}

/**
 * 一次请求可以指名的分支名。它刻意比 git 自己更严：这里没有任何东西需要一个带着
 * 控制字符或者通配符的 ref。
 */
export function validRef(value: string): boolean {
  if (value === "" || value.length > 255) return false;
  if (value.startsWith("-") || value.includes("..")) return false;
  for (const character of value) {
    const code = character.codePointAt(0) ?? 0;
    if (code <= 0x20 || code === 0x7f) return false;
    if ("~^:?*[\\".includes(character)) return false;
  }
  return true;
}

/**
 * 在要一个 PR 之前先确认 head 分支在远端存在，这样一个从没被推上去的分支得到的是
 * 一个清楚的答案，而不是远端一句含糊的拒绝。
 */
export async function createPull(
  service: GithubService,
  caller: Caller,
  request: CreateGithubPullRequest,
): Promise<GithubPullRequest> {
  service.authorize(caller, SCOPE_WRITE);
  const client = service.client();
  const ref = service.repository(request.repository);
  const base = request.baseRef;
  const head = request.headRef;
  if (!validRef(base) || !validRef(head) || base === head) {
    throw githubError("invalid");
  }
  if (
    request.title === "" ||
    Buffer.byteLength(request.title, "utf8") > 256 ||
    Buffer.byteLength(request.body, "utf8") > 65536
  ) {
    throw githubError("invalid");
  }
  let remoteHead: string;
  try {
    remoteHead = await api.refSha(client, ref, head);
  } catch (error) {
    if (codeOf(error) === "NOT_FOUND") throw githubError("notFound");
    throw service.translate(error);
  }
  // 调用方传的是它认为已经推上去的那个 head；从那以后动过的分支会为错误的提交开
  // 一个 PR。
  if (request.expectedHeadSha !== "" && request.expectedHeadSha !== remoteHead) {
    throw githubError("conflict");
  }
  const body: Record<string, unknown> = {
    title: request.title,
    base,
    head,
  };
  let text = request.body;
  if (request.linkedIssueNumber > 0n) {
    // 这条连接被写成远端自己看得懂的文本，同时也在下面被本地记成一条连接。
    if (text !== "") text += "\n\n";
    text += `Closes #${request.linkedIssueNumber}`;
  }
  if (text !== "") body.body = text;
  if (request.draft) body.draft = true;
  const allowed = await mergeMethods(service, client, ref);
  try {
    const pull = await api.createPull(
      client,
      ref,
      body,
      allowed,
      service.now(),
    );
    service.credentials.noteSuccess();
    return pull;
  } catch (error) {
    throw service.translate(error);
  }
}

/**
 * 发一条锚在评审者读过的那个提交上的评审。行内评论带着那个提交，所以一份过时的
 * 草稿是可以认出来的，而不是被悄悄挂到一条它已经不再描述的行上。
 */
export async function submitReview(
  service: GithubService,
  caller: Caller,
  request: SubmitGithubReviewRequest,
): Promise<GithubReview> {
  service.authorize(caller, SCOPE_WRITE);
  const client = service.client();
  const ref = service.repository(request.repository);
  let event: string;
  switch (request.state) {
    case GithubReviewState.APPROVED:
      event = "APPROVE";
      break;
    case GithubReviewState.CHANGES_REQUESTED:
      event = "REQUEST_CHANGES";
      break;
    case GithubReviewState.COMMENTED:
      event = "COMMENT";
      break;
    default:
      throw githubError("invalid");
  }
  if (event === "REQUEST_CHANGES" && request.body.trim() === "") {
    throw githubError("invalid");
  }
  if (
    Buffer.byteLength(request.body, "utf8") > 65536 ||
    request.comments.length > 200
  ) {
    throw githubError("invalid");
  }
  const body: Record<string, unknown> = { event };
  if (request.commitSha !== "") body.commit_id = request.commitSha;
  if (request.body !== "") body.body = request.body;
  const comments: Record<string, unknown>[] = [];
  for (const draft of request.comments) {
    if (
      draft.path === "" ||
      draft.path.length > 4096 ||
      draft.line <= 0n ||
      draft.body.trim() === ""
    ) {
      throw githubError("invalid");
    }
    if (draft.side !== "LEFT" && draft.side !== "RIGHT") {
      throw githubError("invalid");
    }
    comments.push({
      path: draft.path,
      line: Number(draft.line),
      side: draft.side,
      body: draft.body,
    });
  }
  if (comments.length > 0) body.comments = comments;
  try {
    const review = await api.createReview(client, ref, request.number, body);
    service.credentials.noteSuccess();
    return review;
  } catch (error) {
    throw service.translate(error);
  }
}

/**
 * 除非远端仍然持有调用方说它显示过的那个 head 和那个检查汇总，否则拒绝。本地一屏
 * 绿灯不是远端仍会接受这次合并的承诺，所以这里全部重新读一遍，而不是信它。
 */
export async function mergePull(
  service: GithubService,
  caller: Caller,
  request: MergeGithubPullRequest,
): Promise<MergeGithubPullResponse> {
  service.authorize(caller, SCOPE_WRITE);
  const client = service.client();
  const ref = service.repository(request.repository);
  const method = mergeMethodName(request.method);
  if (request.expectedHeadSha === "") throw githubError("invalid");
  const now = service.now();
  const allowed = await mergeMethods(service, client, ref);
  let pull: GithubPullRequest;
  try {
    pull = (await api.pull(client, ref, request.number, allowed, now)).pull;
    service.credentials.noteSuccess();
  } catch (error) {
    throw service.translate(error);
  }
  const refuse = (
    reason: string,
    checks: GithubCheckSummary | undefined,
  ): MergeGithubPullResponse =>
    create(MergeGithubPullResponseSchema, {
      merged: false,
      reasonCode: reason,
      pull,
      ...(checks === undefined ? {} : { checks }),
    });

  if (pull.headSha !== request.expectedHeadSha) {
    return refuse("HEAD_MOVED", undefined);
  }
  if (allowed.length > 0 && !allowed.includes(request.method)) {
    return refuse("METHOD_NOT_ALLOWED", undefined);
  }
  let checks: GithubCheckSummary | undefined;
  if (request.expectedCheckRollup !== GithubCheckConclusion.UNSPECIFIED) {
    try {
      checks = await api.checks(client, ref, pull.headSha, now);
    } catch (error) {
      throw service.translate(error);
    }
    if (checks.rollup !== request.expectedCheckRollup) {
      return refuse("CHECKS_CHANGED", checks);
    }
  }
  if (pull.mergeable === GithubMergeableState.CONFLICTING) {
    return refuse("NOT_MERGEABLE", checks);
  }
  if (pull.mergeable === GithubMergeableState.BLOCKED) {
    // 保护规则和合并队列是远端的决定；core 如实报告被告知的事，而不是硬试一次。
    return refuse("BLOCKED", checks);
  }

  let sha: string;
  try {
    sha = await api.merge(
      client,
      ref,
      pull.number,
      pull.headSha,
      method,
      request.commitTitle,
      request.commitMessage,
    );
  } catch (error) {
    const translated = service.translate(error);
    if (isGithubError(translated, "unknownOutcome")) {
      // 合并可能已经被接受。重新读是唯一诚实的答案；重试会冒出第二个合并提交。
      const latest = await readPull(service, client, ref, pull.number, allowed);
      if (latest !== undefined) {
        pull = latest;
        if (latest.state === GithubPullState.MERGED) {
          return create(MergeGithubPullResponseSchema, {
            merged: true,
            mergeSha: latest.headSha,
            pull: latest,
            ...(checks === undefined ? {} : { checks }),
          });
        }
      }
      return refuse("UNKNOWN_OUTCOME", checks);
    }
    if (
      isGithubError(translated, "conflict") ||
      isGithubError(translated, "invalid")
    ) {
      // 远端拒绝了。是哪一种拒绝由证据决定：head 仍然对得上，意味着被拒的是合并
      // 本身，而不是分支在读者眼皮底下动了。
      const latest = await readPull(service, client, ref, pull.number, allowed);
      if (latest !== undefined) {
        pull = latest;
        if (latest.headSha !== request.expectedHeadSha) {
          return refuse("HEAD_MOVED", checks);
        }
      }
      return refuse("NOT_MERGEABLE", checks);
    }
    if (isGithubError(translated, "permission")) {
      return refuse("BLOCKED", checks);
    }
    throw translated;
  }
  const latest = await readPull(service, client, ref, pull.number, allowed);
  if (latest !== undefined) pull = latest;
  return create(MergeGithubPullResponseSchema, {
    merged: true,
    mergeSha: sha,
    pull,
    ...(checks === undefined ? {} : { checks }),
  });
}

async function readPull(
  service: GithubService,
  client: GithubClient,
  ref: GithubRepositoryRef,
  number: bigint,
  allowed: readonly GithubMergeMethod[],
): Promise<GithubPullRequest | undefined> {
  try {
    return (await api.pull(client, ref, number, allowed, service.now())).pull;
  } catch {
    return undefined;
  }
}

export function mergeMethodName(method: GithubMergeMethod): string {
  switch (method) {
    case GithubMergeMethod.MERGE:
      return "merge";
    case GithubMergeMethod.SQUASH:
      return "squash";
    case GithubMergeMethod.REBASE:
      return "rebase";
    default:
      throw githubError("invalid");
  }
}
